packages/language-server/src/plugins/typescript/features/SemanticTokensProvider.ts

```diff
@@ -0,0 +1,107 @@
+import ts from 'typescript';
+import {
+    Range,
+    SemanticTokens,
+    SemanticTokensBuilder
+} from 'vscode-languageserver';
+import { Document } from '../../../lib/documents';
+import { SemanticTokensProvider } from '../../interfaces';
+import { SnapshotFragment } from '../DocumentSnapshot';
+import { LSAndTSDocResolver } from '../LSAndTSDocResolver';
+import { convertToTextSpan } from '../utils';
+
+export class SemanticTokensProviderImpl implements SemanticTokensProvider {
+    constructor(private readonly lsAndTsDocResolver: LSAndTSDocResolver) {}
+
+    async getSemanticTokens(textDocument: Document, range?: Range): Promise<SemanticTokens> {
+        const { lang, tsDoc } = this.lsAndTsDocResolver.getLSAndTSDoc(textDocument);
+        const fragment = await tsDoc.getFragment();
+        const textSpan = range
+            ? convertToTextSpan(range, fragment)
+            : {
+                  start: 0,
+                  length: fragment.text.length
+              };
```

Comment:
We can save some computation here by ignoring everything that's appended by svelte2tsx which will not be mappable. (Tests would need to be adjusted)

```suggestion
        const textSpan = range
            ? convertToTextSpan(range, fragment)
            : {
                  start: 0,
                  length: tsDoc.parserError
                      ? fragment.text.length
                      : // This is appended by svelte2tsx, there's nothing mappable afterwards
                        fragment.text.lastIndexOf('return { props:') || fragment.text.length
              };
```